File: Python/gamespecs/file.py
import os, pathlib, platform, psutil, winreg
from zipfile import ZipFile
from openstk.poly import Reader, findType
from . import store
from .util import _list

GAMESPATH = 'Games'

# forwards
class FileManager: pass

# tag::FileManager[]
class FileManager:
    class PathItem:
        root: str
        typex: str
        paths: list[str]
        def __init__(self, root: str, elem: dict[str, object]):
            self.root = root
            self.type = _value(elem, 'type') if 'type' in elem else None
            self.paths = _list(elem, 'path') if 'path' in elem else []
        def add(self, root: str, elem: dict[str, object]):
            raise NotImplementedError()

    applicationPath = os.getcwd()
    filters: dict[str, object] = {}
    paths: dict[str, PathItem] = {}
    ignores: dict[str, object] = {}

    # get locale games
    gameRoots = [os.path.join(x.mountpoint, GAMESPATH) for x in psutil.disk_partitions()]
    if platform.system() == 'Android': gameRoots.append(os.path.join('/sdcard', GAMESPATH))
    localGames = {x:os.path.join(r,x) for r in gameRoots if os.path.isdir(r) for x in os.listdir(r)}

    def __init__(self, elem: dict[str, object]):
        # applications
        if 'application' in elem:
            for k,v in elem['application'].items():
                if not k in self.paths:
                    self.addApplication(k, v)
        # direct
        if 'direct' in elem:
            for k,v in elem['direct'].items():
                if 'path' in v:
                    for path in _list(v, 'path'):
                        self.addPath(k, elem, path)
        # ignores
        if 'ignores' in elem:
            for k,v in elem['ignores'].items():
                self.addIgnore(k, _list(v, 'path'))
        # filters
        if 'filters' in elem:
            for k,v in elem['filters'].items():
                self.addFilter(k, v)
    def __repr__(self): return f'''
- paths: {list(self.paths.keys()) if self.paths else None}
- ignores: {list(self.ignores.keys()) if self.ignores else None}
- filters: {list(self.filters.keys()) if self.filters else None}'''

    def merge(self, source: FileManager) -> None:
        self.paths.update(source.paths)
        self.ignores.update(source.ignores)
        self.filters.update(source.filters)

    def addApplication(self, id: str, elem: dict[str, object]) -> None:
        if platform.system() == 'Windows' and 'reg' in elem:
            for k in _list(elem, 'reg'):
                if not id in self.paths and (z := self.getPathByRegistryKey(k, elem[k] if k in elem else None)):
                    self.addPath(id, elem, z)
        if 'key' in elem:
            for k in _list(elem, 'key'):
                if not id in self.paths and (z := store.getPathByKey(k)):
                    self.addPath(id, elem, z)
        if 'dir' in elem:
            for k in _list(elem, 'dir'):
                if not id in self.paths and k in FileManager.localGames:
                    self.addPath(id, elem, FileManager.localGames[k])
    
    def addFilter(self, id, elem: dict[str, object]) -> None:
        if not id in self.filters: self.filters[id] = {}
        z2 = self.filters[id]
        for k, v in elem.items():
            z2[k] = v

    def addIgnore(self, id: str, paths: list[str]) -> None:
        if not id in self.ignores: self.ignores[id] = set()
        z2 = self.ignores[id]
        for v in paths: z2.add(v)

    def addPath(self, id: str, elem: dict[str, object], path: str) -> None:
        if not path: raise Exception('Require Path')
        path = FileManager.getPathWithSpecialFolders(path, '')
        if not os.path.isdir(path) and not os.path.isfile(path): return
        if not id in self.paths: self.paths[id] = self.PathItem(path, elem)
        else: self.paths[id].add(path, elem)

    # tag::FileManager.getPathWithSpecialFolders[]
    @staticmethod
    def getPathWithSpecialFolders(path: str, rootPath: str = None) -> str:
        lowerPath = path.lower()
        return f'{rootPath}{path[6:]}' if lowerPath.startswith('%path%') else \
        f'{FileManager.applicationPath}{path[9:]}' if lowerPath.startswith('%apppath%') else \
        f'{os.getenv("APPDATA")}{path[9:]}' if lowerPath.startswith('%appdata%') else \
        f'{os.getenv("LOCALAPPDATA")}{path[14:]}' if lowerPath.startswith('%localappdata%') else \
        path
    # end::FileManager.getPathWithSpecialFolders[]

    # tag::FileManager.findRegistryPath[]
    @staticmethod
    def findRegistryPath(paths: list[str]) -> str:
        for p in paths:
            keyPath = p.replace('/', '\\')
            try: key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, f'SOFTWARE\\{keyPath}', 0, winreg.KEY_READ)
            except FileNotFoundError:
                try: key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, f'SOFTWARE\\{keyPath}', 0, winreg.KEY_READ)
                except FileNotFoundError:
                    try: key = winreg.OpenKey(winreg.HKEY_CLASSES_ROOT, f'VirtualStore\\MACHINE\\SOFTWARE\\{keyPath}', 0, winreg.KEY_READ)
                    except FileNotFoundError: key = None
            if key is None: continue
            # search directories
            path = None
            for search in ['Path', 'Install Dir', 'InstallDir', 'InstallLocation']:
                try:
                    val = winreg.QueryValueEx(key, search)[0]
                    if os.path.isdir(val): path = val; break
                except FileNotFoundError: continue
            # search files
            if path is None:
                for search in ['Installed Path', 'ExePath', 'Exe']:
                    try:
                        val = winreg.QueryValueEx(key, search)[0]
                        if os.path.exists(val): path = val; break
                    except FileNotFoundError: continue
                if path is not None: path = os.path.dirname(path)
            if path is not None and os.path.isdir(path): return path
        return None
    # end::FileManager.findRegistryPath[]

    @staticmethod
    def getPathByRegistryKey(key: str, elem: dict[str, object]) -> str:
        path = FileManager.findRegistryPath([f'Wow6432Node\\{key}', key])
        if not elem: return path
        elif 'path' in elem: return os.path.abspath(FileManager.getPathWithSpecialFolders(elem['path'], path))
        elif 'xml' in elem and 'xmlPath' in elem:
            return FileManager.getSingleFileValue(FileManager.getPathWithSpecialFolders(elem['xml'], path), 'xml', elem['xmlPath'])
        return None

    @staticmethod
    def getSingleFileValue(path: str, ext: str, select: str) -> str:
        if not os.fileExists(path): return None
        with open(path, 'r') as f: content = f.read()
        match ext:
            case 'xml': raise NotImplementedError() #return XDocument.Parse(content).XPathSelectElement(select)?.Value,
            case _: raise Exception(f'Unknown {ext}')
        return os.path.basename(value) if value else None
# end::FileManager[]

# tag::FileSystem[]

# IFileSystem
class IFileSystem:
    def glob(self, path: str, searchPattern: str) -> list[str]: pass
    def fileExists(self, path: str) -> bool: pass
    def fileInfo(self, path: str) -> (str, int): pass
    def openReader(self, path: str, mode: str = 'rb') -> Reader: pass
    def findPaths(self, path: str, searchPattern: str) -> str:
        if (expandStartIdx := searchPattern.find('(')) != -1 and \
            (expandMidIdx := searchPattern.find(':', expandStartIdx)) != -1 and \
            (expandEndIdx := searchPattern.find(')', expandMidIdx)) != -1 and \
            expandStartIdx < expandEndIdx:
            for expand in searchPattern[expandStartIdx + 1: expandEndIdx].split(':'):
                for found in self.findPaths(path, searchPattern[:expandStartIdx] + expand + searchPattern[expandEndIdx+1:]): yield found
            return
        for path in self.glob(path, searchPattern): yield path

# StandardFileSystem
class StandardFileSystem(IFileSystem):
    def __init__(self, root: str): self.root = root; self.skip = len(root)
    def glob(self, path: str, searchPattern: str) -> list[str]:
        g = pathlib.Path(os.path.join(self.root, path)).glob(searchPattern if searchPattern else '**/*')
        return [str(x)[self.skip:] for x in g]
    def fileExists(self, path: str) -> bool: return os.path.exists(os.path.join(self.root, path))
    def fileInfo(self, path: str) -> (str, int): return (path, os.stat(path).st_size) if os.path.exists(path := os.path.join(self.root, path)) else (None, 0)
    def openReader(self, path: str, mode: str = 'rb') -> Reader: return Reader(open(os.path.join(self.root, path), mode))

# ZipFileSystem
class ZipFileSystem(IFileSystem):
    def __init__(self, root: str, path: str):
        self.pak = ZipFile(root)
        self.root = '' if not root else f'{root}/'
    def glob(self, path: str, searchPattern: str) -> list[str]:
        root = os.path.join(self.root, path)
        skip = len(root)
        return []
    def fileExists(self, path: str) -> bool: return self.pak.read(os.path.join(self.root, path)) != None
    def fileInfo(self, path: str) -> (str, int): e = self.pak.read(os.path.join(self.root, path)); return (e.name, e.length) if e else (None, 0)
    def openReader(self, path: str, mode: str = 'rb') -> Reader: return Reader(self.pak.read(os.path.join(self.root, path)))

# ZipIsoFileSystem
class ZipIsoFileSystem(IFileSystem):
    def __init__(self, root: str, path: str):
        self.pak = ZipFile(root)
        self.root = '' if not root else f'{root}/'
    def glob(self, path: str, searchPattern: str) -> list[str]:
        root = os.path.join(self.root, path)
        skip = len(root)
        return []
    def fileExists(self, path: str) -> bool: return self.pak.read(path) != None
    def fileInfo(self, path: str) -> (str, int): e = self.pak.read(path); return (e.name, e.length) if e else (None, 0)
    def openReader(self, path: str, mode: str = 'rb') -> Reader: return Reader(self.pak.read(path))

# HostFileSystem
class HostFileSystem(IFileSystem):
    def __init__(self, uri: str): self.uri = uri
    def glob(self, path: str, searchPattern: str) -> list[str]: raise NotImplementedError()
    def fileExists(self, path: str) -> bool: raise NotImplementedError()
    def fileInfo(self, path: str) -> (str, int): raise NotImplementedError()
    def openReader(self, path: str, mode: str = 'rb') -> Reader: raise NotImplementedError()

# end::FileSystem[]

File: dotnet/Base/GameX.Base/File.cs
using GameX.Formats;
using GameX.Platforms;
using Microsoft.Extensions.FileSystemGlobbing;
using Microsoft.Win32;
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Xml.Linq;
using System.Xml.XPath;
using static GameX.Util;
using static Microsoft.Win32.Registry;

namespace GameX
{
    #region FileManager

    /// <summary>
    /// FileManager
    /// </summary>
    public class FileManager
    {
        const string GAMESPATH = "Games";

        public class PathItem
        {
            public string Root;
            public string Type;
            public string[] Paths;

            public PathItem(string root, JsonElement elem)
            {
                Root = root;
                Type = elem.TryGetProperty("type", out var z) ? z.GetString() : null;
                Paths = elem.TryGetProperty("path", out z) ? _listV(z) : new string[0];
            }

            public void Add(string root, JsonElement elem)
                => throw new NotSupportedException();
        }

        /// <summary>
        /// Gets the host factory.
        /// </summary>
        /// <value>
        /// The host factory.
        /// </value>
        public virtual Func<Uri, string, AbstractHost> HostFactory { get; } = HttpHost.Factory;

        /// <summary>
        /// Gets a value indicating whether this instance has locations.
        /// </summary>
        /// <value>
        ///   <c>true</c> if this instance is data present; otherwise, <c>false</c>.
        /// </value>
        public bool HasPaths => Paths.Count != 0;

        /// <summary>
        /// The locations
        /// </summary>
        public readonly IDictionary<string, PathItem> Paths = new Dictionary<string, PathItem>();

        /// <summary>
        /// The ignores
        /// </summary>
        public readonly IDictionary<string, HashSet<string>> Ignores = new Dictionary<string, HashSet<string>>();

        /// <summary>
        /// The filters
        /// </summary>
        public readonly IDictionary<string, IDictionary<string, string>> Filters = new Dictionary<string, IDictionary<string, string>>();

        public static readonly IDictionary<string, string> LocalGames;

        static FileManager()
        {
            // get locale games
            var gameRoots = DriveInfo.GetDrives().Select(x => Path.Combine(x.Name, GAMESPATH)).ToList();
            if (Platform.PlatformOS == Platform.OS.Android) gameRoots.Add(Path.Combine("/sdcard", GAMESPATH));
            LocalGames = gameRoots.Where(Directory.Exists).SelectMany(Directory.GetDirectories).ToDictionary(Path.GetFileName, x => x);
        }

        internal FileManager() { }
        public FileManager(JsonElement elem)
        {
            // applications
            if (elem.TryGetProperty("application", out var z))
                foreach (var prop in z.EnumerateObject())
                    if (!Paths.ContainsKey(prop.Name))
                        AddApplication(prop.Name, prop.Value);
            // direct
            if (elem.TryGetProperty("direct", out z))
                foreach (var prop in z.EnumerateObject())
                    if (prop.Value.TryGetProperty("path", out z))
                        foreach (var path in _listV(z))
                            AddPath(prop.Name, prop.Value, path);
            // ignores
            if (elem.TryGetProperty("ignores", out z))
                foreach (var prop in z.EnumerateObject())
                    AddIgnore(prop.Name, _list(prop.Value, "path"));
            // filters
            if (elem.TryGetProperty("filters", out z))
                foreach (var prop in z.EnumerateObject())
                    AddFilter(prop.Name, prop.Value);
        }

        /// <summary>
        /// Merges the family.
        /// </summary>
        /// <param name="source">The source.</param>
        public void Merge(FileManager source)
        {
            foreach (var s in source.Paths) Paths.Add(s.Key, s.Value);
            foreach (var s in source.Ignores) Ignores.Add(s.Key, s.Value);
            foreach (var s in source.Filters) Filters.Add(s.Key, s.Value);
        }

        #region Parse File-Manager

        protected void AddApplication(string id, JsonElement elem)
        {
            string z;
            if (Platform.PlatformOS == Platform.OS.Windows && elem.TryGetProperty("reg", out var y))
                foreach (var k in _listV(y))
                    if (!Paths.ContainsKey(id) && (z = GetPathByRegistryKey(k, elem.TryGetProperty(k, out y) ? y : (JsonElement?)null)) != null)
                        AddPath(id, elem, z);
            if (elem.TryGetProperty("key", out y))
                foreach (var k in _listV(y))
                    if (!Paths.ContainsKey(id) && (z = StoreManager.GetPathByKey(k)) != null)
                        AddPath(id, elem, z);
            if (elem.TryGetProperty("dir", out y))
                foreach (var k in _listV(y))
                    if (!Paths.ContainsKey(id) && LocalGames.TryGetValue(k, out var path))
                        AddPath(id, elem, path);
        }

        protected void AddFilter(string id, JsonElement elem)
        {
            if (!Filters.TryGetValue(id, out var z2)) Filters.Add(id, z2 = new Dictionary<string, string>());
            foreach (var z in elem.EnumerateObject())
                z2.Add(z.Name, _valueV(z.Value).ToString());
        }

        protected void AddIgnore(string id, string[] paths)
        {
            if (!Ignores.TryGetValue(id, out var z2)) Ignores.Add(id, z2 = new HashSet<string>());
            foreach (var v in paths) z2.Add(v);
        }

        protected void AddPath(string id, JsonElement elem, string path)
        {
            if (path == null) throw new ArgumentOutOfRangeException(nameof(path));
            path = Path.GetFullPath(GetPathWithSpecialFolders(path));
            if (!Directory.Exists(path) && !File.Exists(path)) return;
            if (!Paths.TryGetValue(id, out var z2)) Paths.Add(id, new PathItem(path, elem));
            else z2.Add(path, elem);
        }

        protected static string GetPathWithSpecialFolders(string path, string rootPath = null) =>
            path.StartsWith("%Path%", StringComparison.OrdinalIgnoreCase) ? $"{rootPath}{path[6..]}"
            : path.StartsWith("%AppPath%", StringComparison.OrdinalIgnoreCase) ? $"{FamilyManager.ApplicationPath}{path[9..]}"
            : path.StartsWith("%AppData%", StringComparison.OrdinalIgnoreCase) ? $"{Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)}{path[9..]}"
            : path.StartsWith("%LocalAppData%", StringComparison.OrdinalIgnoreCase) ? $"{Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData)}{path[14..]}"
            : path;

        /// <summary>
        /// Gets the executable path.
        /// </summary>
        /// <param name="name">Name of the sub.</param>
        /// <returns></returns>
        protected static string FindRegistryPath(string[] paths)
        {
            var localMachine64 = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64);
            var currentUser64 = RegistryKey.OpenBaseKey(RegistryHive.CurrentUser, RegistryView.Registry64);
            foreach (var p in paths)
                try
                {
                    var keyPath = p.Replace('/', '\\');
                    var key = new Func<RegistryKey>[] {
                        () => localMachine64.OpenSubKey($"SOFTWARE\\{keyPath}"),
                        () => currentUser64.OpenSubKey($"SOFTWARE\\{keyPath}"),
                        () => ClassesRoot.OpenSubKey($"VirtualStore\\MACHINE\\SOFTWARE\\{keyPath}") }
                        .Select(x => x()).FirstOrDefault(x => x != null);
                    if (key == null) continue;
                    // search directories
                    var path = new[] { "Path", "Install Dir", "InstallDir", "InstallLocation" }
                        .Select(x => key.GetValue(x) as string)
                        .FirstOrDefault(x => !string.IsNullOrEmpty(x) && Directory.Exists(x));
                    if (path == null)
                    {
                        // search files
                        path = new[] { "Installed Path", "ExePath", "Exe" }
                            .Select(x => key.GetValue(x) as string)
                            .FirstOrDefault(x => !string.IsNullOrEmpty(x) && File.Exists(x));
                        if (path != null) path = Path.GetDirectoryName(path);
                    }
                    if (path != null && Directory.Exists(path)) return path;
                }
                catch { return null; }
            return null;
        }

        protected static string GetPathByRegistryKey(string key, JsonElement? elem)
        {
            var path = FindRegistryPath(new[] { $@"Wow6432Node\{key}", key });
            if (elem == null) return path;
            else if (elem.Value.TryGetProperty("path", out var z)) return Path.GetFullPath(GetPathWithSpecialFolders(z.GetString(), path));
            else if (elem.Value.TryGetProperty("xml", out z) && elem.Value.TryGetProperty("xmlPath", out var y))
                return GetSingleFileValue(GetPathWithSpecialFolders(z.GetString(), path), "xml", y.GetString());
            else return null;
        }

        protected static string GetSingleFileValue(string path, string ext, string select)
        {
            if (!File.Exists(path)) return null;
            var content = File.ReadAllText(path);
            var value = ext switch
            {
                "xml" => XDocument.Parse(content).XPathSelectElement(select)?.Value,
                _ => throw new ArgumentOutOfRangeException(nameof(ext)),
            };
            return value != null ? Path.GetDirectoryName(value) : null;
        }

        #endregion
    }

    #endregion

    #region IFileSystem

    /// <summary>
    /// IFileSystem
    /// </summary>
    public interface IFileSystem
    {
        IEnumerable<string> Glob(string path, string searchPattern);
        bool FileExists(string path);
        (string path, long length) FileInfo(string path);
        BinaryReader OpenReader(string path);
        BinaryWriter OpenWriter(string path);
    }

    #endregion

    #region StandardFileSystem

    /// <summary>
    /// StandardFileSystem
    /// </summary>
    internal class StandardFileSystem : IFileSystem
    {
        readonly string Root;
        readonly int Skip;
        public StandardFileSystem(string root) { Root = root; Skip = Root.Length + 1; }
        public IEnumerable<string> Glob(string path, string searchPattern)
        {
            var matcher = new Matcher();
            matcher.AddIncludePatterns(new[] { string.IsNullOrEmpty(searchPattern) ? "**/*" : searchPattern });
            return matcher.GetResultsInFullPath(Path.Combine(Root, path)).Select(x => x[Skip..]);
        }
        public bool FileExists(string path) => File.Exists(Path.Combine(Root, path));
        public (string path, long length) FileInfo(string path) => File.Exists(path = Path.Combine(Root, path)) ? (path[Skip..], new FileInfo(Path.Combine(Root, path)).Length) : (null, 0);
        public BinaryReader OpenReader(string path) => new BinaryReader(File.Open(Path.Combine(Root, path), FileMode.Open, FileAccess.Read, FileShare.Read));
        public BinaryWriter OpenWriter(string path) => new BinaryWriter(File.Open(Path.Combine(Root, path), FileMode.Open, FileAccess.Write, FileShare.Write));

        public static Func<string, bool> CreateMatcher(string searchPattern)
        {
            if (string.IsNullOrEmpty(searchPattern)) return x => true;
            var wildcardCount = searchPattern.Count(x => x.Equals('*'));
            if (wildcardCount <= 0) return x => x.Equals(searchPattern, StringComparison.CurrentCultureIgnoreCase);
            else if (wildcardCount == 1)
            {
                var newPattern = searchPattern.Replace("*", "");
                if (searchPattern.StartsWith("*")) return x => x.EndsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
                else if (searchPattern.EndsWith("*")) return x => x.StartsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
            }
            var regexPattern = string.Concat("^", Regex.Escape(searchPattern).Replace("\\*", ".*"), "$");
            return x =>
            {
                try { return Regex.IsMatch(x, regexPattern); }
                catch { return false; }
            };
        }
    }

    #endregion

    #region ZipFileSystem

    /// <summary>
    /// ZipFileSystem
    /// </summary>
    internal class ZipFileSystem : IFileSystem
    {
        readonly ZipArchive Pak;
        readonly string Root;
        public ZipFileSystem(string root, string path)
        {
            Pak = ZipFile.Open(root, ZipArchiveMode.Read);
            Root = string.IsNullOrEmpty(path) ? string.Empty : $"{path}{Path.AltDirectorySeparatorChar}";
        }
        public IEnumerable<string> Glob(string path, string searchPattern)
        {
            var root = Path.Combine(Root, path);
            var skip = root.Length;
            var matcher = StandardFileSystem.CreateMatcher(searchPattern);
            return Pak.Entries.Where(x =>
            {
                var fn = x.FullName;
                return fn.Length > skip && fn.StartsWith(root) && matcher(fn[skip..]);
            }).Select(x => x.FullName[skip..]);
        }
        public bool FileExists(string path) => Pak.GetEntry(Path.Combine(Root, path)) != null;
        public (string path, long length) FileInfo(string path) { var e = Pak.GetEntry(Path.Combine(Root, path)); return e != null ? (e.Name, e.Length) : (null, 0); }
        public BinaryReader OpenReader(string path) => new BinaryReader(Pak.GetEntry(Path.Combine(Root, path)).Open());
        public BinaryWriter OpenWriter(string path) => throw new NotSupportedException();
    }

    //public static Func<string, bool> CreateMatcher(string root, string searchPattern)
    //{
    //    if (string.IsNullOrEmpty(root)) return CreateMatcher(searchPattern);
    //    var matcher = CreateMatcher(searchPattern);
    //    var skip = root.Length;
    //    return x => x.StartsWith(root) && matcher(x[skip..]);
    //}

    #endregion

    #region ZipIsoFileSystem

    /// <summary>
    /// ZipIsoFileSystem
    /// </summary>
    internal class ZipIsoFileSystem : IFileSystem
    {
        readonly ZipArchive Pak;
        readonly string Path;
        public ZipIsoFileSystem(string root, string path)
        {
            Pak = ZipFile.Open(root, ZipArchiveMode.Read);
            Path = path;
        }
        public IEnumerable<string> Glob(string path, string searchPattern)
        {
            var matcher = StandardFileSystem.CreateMatcher(searchPattern);
            return Pak.Entries.Where(x => matcher(x.Name)).Select(x => x.Name);
        }
        public bool FileExists(string path) => Pak.GetEntry(path) != null;
        public (string path, long length) FileInfo(string path) { var e = Pak.GetEntry(path); return e != null ? (e.Name, e.Length) : (null, 0); }
        public BinaryReader OpenReader(string path) => new BinaryReader(Pak.GetEntry(path).Open());
        public BinaryWriter OpenWriter(string path) => throw new NotSupportedException();
    }

    #endregion

    #region HostFileSystem

    /// <summary>
    /// HostFileSystem
    /// </summary>
    internal class HostFileSystem : IFileSystem
    {
        public HostFileSystem(Uri uri)
        {
            if (uri == null) throw new ArgumentNullException(nameof(uri));
            var pathOrPattern = uri.LocalPath;
            var searchPattern = Path.GetFileName(pathOrPattern);
            var path = Path.GetDirectoryName(pathOrPattern);
            // file
            if (!string.IsNullOrEmpty(searchPattern)) throw new ArgumentOutOfRangeException(nameof(pathOrPattern), pathOrPattern); //: Web single file access to supported.

            //options = PakOption.Stream;
            //searchPattern = Path.GetFileName(path);
            //path = Path.GetDirectoryName(path);
            //if (path.Contains('*')) throw new NotSupportedException("Web wildcard folder access");
            //host = new UriBuilder(uri) { Path = $"{path}/", Fragment = null }.Uri;
            //if (searchPattern.Contains('*'))
            //{
            //    var set = new HttpHost(host).GetSetAsync().Result ?? throw new NotSupportedException(".set not found. Web wildcard access");
            //    var pattern = $"^{Regex.Escape(searchPattern.Replace('*', '%')).Replace("_", ".").Replace("%", ".*")}$";
            //    return set.Where(x => Regex.IsMatch(x, pattern)).ToArray();
            //}
            //return new[] { searchPattern };
        }
        public IEnumerable<string> Glob(string path, string searchPattern)
        {
            var matcher = new Matcher();
            matcher.AddIncludePatterns(new[] { searchPattern });
            return matcher.GetResultsInFullPath(searchPattern);
        }
        public bool FileExists(string path) => File.Exists(path);
        public (string path, long length) FileInfo(string path) => File.Exists(path) ? (path, 0) : (null, 0);
        public BinaryReader OpenReader(string path) => null;
        public BinaryWriter OpenWriter(string path) => null;
    }

    #endregion
}